xtm/xtm_api.cc

```diff
@@ -0,0 +1,353 @@
+ * SUCH DAMAGE.
+ */
+#include "xtm_api.h"
+#include "xtm_scsp_queue.h"
+#include "xtm_config.h"
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <assert.h>
+#include <fcntl.h>
+#include <errno.h>
+#ifdef TARANTOOL_XTM_USE_EVENTFD
+#include <sys/eventfd.h>
+#endif /* defined(TARANTOOL_XTM_USE_EVENTFD) */
+
+#define XTM_PIPE_SIZE 4096
+
+union xtm_msg {
+	/**
+	 * Anonymous structure for storing functions and their arguments,
+	 * when using the xtm API in dispatch/invoke pattern.
+	 */
+	struct {
+		xtm_queue_fun_t fun;
+		void *fun_arg;
+	};
+	/**
+	 * Pointer to save message, when using xtm API
+	 * in send/recv pattern.
+	 */
+	void *ptr;
+};
+
+struct xtm_queue {
+	/**
+	 * File descriptor that the consumer thread must poll,
+	 * to know when new messages are added to the queue.
+	 */
+	int consumer_read_fd;
+	/**
+	 * File descriptor to which the producer thread writes
+	 * to notify the consumer thread of new messages in the queue.
+	 */
+	int consumer_write_fd;
+	/**
+	 * File descriptor that the producer thread must poll,
+	 * to know when there is a free space in qeuue.
+	 */
+	int producer_read_fd;
+	/**
+	 * File descriptor to which the consumer thread writes
+	 * to notify the producer thread about free space in queue.
+	 */
+	int producer_write_fd;
+	/**
+	 * Flag indicates, that producer, wants to get notification
+	 * about free place in queue.
+	 */
+	bool is_producer_need_notifications;
+	/**
+	 * Flag indicates, that we need to close all read file descriptors
+	 * manually, when we delete xtm queue.
+	 */
+	bool is_need_to_close_readfd;
+	/**
+	 * Flag indicates, that producer couldn't put an item in
+	 * the queue, and wait for notification.
+	 */
+	bool is_notification_expected;
+	/** Message queue, it's size must be power of two */
+	struct xtm_scsp_queue<union xtm_msg> queue;
+};
+
+static inline int
+notify_fd(int fd)
+{
+	uint64_t tmp = 1;
+	ssize_t cnt;
+	/*
+	 * We must write 8 byte value, because for linux we
+	 * used eventfd, which require to write >= 8 byte at once.
+	 * Also in case of EINTR we retry to write
+	 */
+	while ((cnt = write(fd, &tmp, sizeof(tmp))) < 0 && errno == EINTR)
+		;
+	return (cnt == sizeof(tmp) ? 0 : -1);
+}
+
+struct xtm_queue *
+xtm_queue_new(unsigned size, bool is_need_to_close_readfd,
+	      bool is_producer_need_notifications)
+{
+	int save_errno = 0;
+	int consumer_filedes[2];
+	int producer_filedes[2];
+	struct xtm_queue *queue = (struct xtm_queue *)
+		malloc(sizeof(struct xtm_queue) +
+		       size * sizeof(union xtm_msg));
+	if (queue == NULL)
+		return NULL;
+
+#ifdef TARANTOOL_XTM_USE_EVENTFD
+	if ((consumer_filedes[0] = eventfd(0, 0)) < 0) {
+#else /* !defined(TARANTOOL_XTM_USE_EVENTFD) */
+	if (pipe(consumer_filedes) < 0) {
+#endif /* defined(TARANTOOL_XTM_USE_EVENTFD) */
+		save_errno = errno;
+		goto free_queue;
+	}
+
+#ifdef TARANTOOL_XTM_USE_EVENTFD
+	queue->consumer_read_fd = consumer_filedes[0];
+	queue->consumer_write_fd = consumer_filedes[0];
```

Comment:
"queue->consumer_read_fd = queue->consumer_write_fd = consumer_filedes[0];"
is shorter and more readable

Reply:
Fixed

Reply:
Both versions are fine to me. It is not specified in the code style how should we do. Only as long as it fits 80 symbols, of course. If it does not fit, you need to have 2 lines.